consensus/bor/bor.go

```diff
@@ -1196,24 +1339,127 @@ func (c *Bor) FetchAndCommitSpan(
 			producers = append(producers, m)
 		}
 	} else {
-		response, err := c.HeimdallClient.GetSpan(ctx, newSpanID)
-		if err != nil {
-			return err
+		getSpanLength := func(startBlock, endBlock uint64) uint64 {
+			return endBlock - startBlock
 		}
-
-		minSpan = span.Span{
-			ID:         response.Id,
-			StartBlock: response.StartBlock,
-			EndBlock:   response.EndBlock,
+		getSpanStartBlock := func() uint64 {
+			return header.Number.Uint64() + c.config.CalculateSprint(header.Number.Uint64())
+		}
+		getSpanEndBlock := func(startBlock uint64, spanLength uint64) uint64 {
+			return startBlock + spanLength
 		}
-		chainId = response.BorChainId
+		retryWithBackoff := func(exec func() (bool, error)) error {
+			i := 1
+			for {
+				done, err := exec()
+				if err != nil {
+					return err
+				}
+
+				if done {
+					return nil
+				}
 
-		for _, val := range response.ValidatorSet.Validators {
-			validators = append(validators, val.MinimalVal())
+				time.Sleep(time.Duration(i) * time.Second)
+
+				if i < 5 {
+					i++
+				}
+			}
 		}
 
-		for _, val := range response.SelectedProducers {
-			producers = append(producers, val.MinimalVal())
+		if hmm.IsHeimdallV2 {
+			var response *borTypes.Span
+			var err error
+
+			err = retryWithBackoff(func() (bool, error) {
+				response, err = c.HeimdallClient.GetSpanV2(ctx, newSpanID)
+				if err == nil {
+					return true, nil
+				}
+				log.Error("Error while fetching heimdallv2 span", "error", err)
+				if response, err = c.getLatestHeimdallSpanV2(); err != nil {
+					log.Error("Error while fetching last heimdallv2 span from db", "error", err)
+					return false, err
+				}
+				if response != nil {
+					response.Id = newSpanID
+					spanLength := getSpanLength(response.StartBlock, response.EndBlock)
+					response.StartBlock = getSpanStartBlock()
+					response.EndBlock = getSpanEndBlock(response.StartBlock, spanLength)
+					return true, nil
```

Comment:
It seems like the startblock and endblock are overridden here for every span. What will happen when heimdall just proposes a new span, will its values get overridden?

Reply:
If heimdall is up this logic will not trigger, this logic triggers just when heimdall is down and we have saved latest heimdall span with the `saveLatestHeimdallSpanVX` call above. We need to update that span id, start block, end block and use same validators and producers set.

Reply:
One thing I failed to understand is why don't we move this update logic to `saveLatestHeimdallSpanVX`, and here we can simply get latest span? In that way, there will be only one place where we know the span could be modified/updated.

Reply:
This logic here will trigger only at last sprint of the span, we want to fetch the latest heimdall span as soon as possible because we dont know when heimdall could go down.